p4-16/spec/P4-16-spec.mdk

```diff
@@ -4971,6 +4973,40 @@ void packet_in.extract<T>(out T headerLValue) {
 }
 ~ End P4Pseudo
 
+The semantics of the single-argument `extract` method on a struct type
+is given below. This use of `extract` is only required to produce
+predictable results if it is performed at the same offset from the
+beginning of the packet that the same target device earlier performed
+an `emit` method on a struct with the same type name. In this case,
+the resulting value of `structLValue` will be equal to the original
+struct value that was emitted, according to the `==` operator. If such
+an `extract` operation is done in any other situation, the resulting
+value of `structLValue` is unspecified.
+
+The length in bits of the data consumed by such an `extract` operation
```

Comment:
I don't like this.
Packets are expected to be exchanged between machines. What's the point of emitting a packet if you don't know how it will look? If we want extract/emit for internal packets only we should use a different extern.

Reply:
Do you mean a different extern object that is neither packet_in nor packet_out?

Or do you mean a different method defined on those objects, with different names that are neither emit nor extract?  @hanw Do you care what the method names are here?  e.g. what if it were pkt.serialize(my_struct) instead of pkt.emit(my_struct), and pkt.deserialize(my_struct) instead of pkt.extract(my_struct)?

Reply:
Extending packet_in and packet_out with serialize() and deserialize() methods sounds like a good idea to me.

Reply:
No, it should be a different extern.
You can call it internal_buffer.
It can have an API which is the union of packet_in and packet_out.

Reply:
While I was re-reading the P4-16 spec again last week, a question came to mind: why do we need packet_in and packet_out extern at all? Why don't we define emit() and extract() to be function that operates on header, as P4-14 did? 

For instance, it seems I could rewrite a P4-16 program with packet_in extern with a simple extract method.

```
parser p (packet_in pkt, header hdr) {
  state start {
     pkt.extract(hdr.ethernet);
  }
}

// can be rewritten as
parser p (header hdr) {
   state start {
      extract(hdr.ethernet);
   }
}

```

What does pass packet_in as a direction-less extern to parser buy us?

Reply:
Having a separate extern may conflict with @hanw's goals, but I will let him speak to that.  Then again, it may simplify things somewhat, since then all deserialized structs can go to "the same place" or "next to each other" in an implementation, yet "separate from" or "away from" the packet contents created via emit(my_header) calls.

Reply:
Han, to maybe clarify my previous comment, if there were a separate extern object on which serialize/deserialize calls were made, vs. on the packet_in/out objects, then there is no need to require an implementation to deal with deserialized struct data "between two emitted headers".

Reply:
Explicit is better than implicit.
The language is simpler, extern is not a keyword, it's a method.
Also, you can possibly handle multiple packet_in objects if your architecture exposes them.

Reply:
Regarding the multiple packet_in objects in the same parser idea, if packet_in is used to represent a stream of bytes that belong to the same packet, then multiple packet_in objects means multiple streams of bytes that belong to different packets. Essentially, it represents a parser block that is connected to multiple ports. 

How do we implement the parse state machine, if there is only one start state, and how do we implement the arbitration logic between different ports?

Reply:
For packet_out there is no issue.
For packet_in the architecture should provide information about the packet state, and it should specify the rules for invoking a parser (or some other block that has packet_in arguments). You could for example use the length of a packet to figure out whether it's present or not.